lightning/src/ln/chan_utils.rs

```diff
@@ -691,58 +776,425 @@ impl HolderCommitmentTransaction {
+
+		let (obscured_commitment_transaction_number, txins) = Self::internal_build_inputs(commitment_number, channel_parameters);
+		let transaction = Self::make_transaction(obscured_commitment_transaction_number, txins, outputs);
+		let txid = transaction.txid();
+		CommitmentTransaction {
+			commitment_number,
+			to_broadcaster_value_sat,
+			to_countersignatory_value_sat,
+			feerate_per_kw,
+			htlcs,
+			keys,
+			built: BuiltCommitmentTransaction {
+				transaction,
+				txid
+			},
+		}
+	}
+
+	#[cfg(feature = "_test_utils")]
+	/// Get the transaction output scripts.
+	///
+	/// For debugging purposes, since only the hash is present in the actual transaction.
+	pub fn get_output_scripts(&self, keys: &TxCreationKeys, channel_parameters: &DirectedChannelTransactionParameters) -> Result<Vec<Script>, ()> {
+		let mut htlcs_with_aux = self.htlcs.iter().map(|h| (h.clone(), ())).collect();
+		let (_, _, scripts) = Self::internal_build_outputs(keys, self.to_broadcaster_value_sat, self.to_countersignatory_value_sat, &mut htlcs_with_aux, channel_parameters)?;
+		Ok(scripts)
+	}
+
+	fn internal_rebuild_transaction(&self, keys: &TxCreationKeys, channel_parameters: &DirectedChannelTransactionParameters) -> Result<BuiltCommitmentTransaction, ()> {
+		let (obscured_commitment_transaction_number, txins) = Self::internal_build_inputs(self.commitment_number, channel_parameters);
+
+		let mut htlcs_with_aux = self.htlcs.iter().map(|h| (h.clone(), ())).collect();
+		let (outputs, _, _) = Self::internal_build_outputs(keys, self.to_broadcaster_value_sat, self.to_countersignatory_value_sat, &mut htlcs_with_aux, channel_parameters)?;
+
+		let transaction = Self::make_transaction(obscured_commitment_transaction_number, txins, outputs);
+		let txid = transaction.txid();
+		let built_transaction = BuiltCommitmentTransaction {
+			transaction,
+			txid
+		};
+		Ok(built_transaction)
+	}
+
+	fn make_transaction(obscured_commitment_transaction_number: u64, txins: Vec<TxIn>, outputs: Vec<TxOut>) -> Transaction {
+		Transaction {
+			version: 2,
+			lock_time: ((0x20 as u32) << 8 * 3) | ((obscured_commitment_transaction_number & 0xffffffu64) as u32),
+			input: txins,
+			output: outputs,
+		}
+	}
+
+	// This is used in two cases:
+	// - initial sorting of outputs / HTLCs in the constructor, in which case T is auxiliary data the
+	//   caller needs to have sorted together with the HTLCs so it can keep track of the output index
+	// - building of a bitcoin transaction during a verify() call, in which case T is just ()
+	fn internal_build_outputs<T>(keys: &TxCreationKeys, to_broadcaster_value_sat: u64, to_countersignatory_value_sat: u64, htlcs_with_aux: &mut Vec<(HTLCOutputInCommitment, T)>, channel_parameters: &DirectedChannelTransactionParameters) -> Result<(Vec<TxOut>, Vec<HTLCOutputInCommitment>, Vec<Script>), ()> {
```

Comment:
Does this actually do anything?  isn't the compiler smart enough to inline if it's appropriate?  also, isn't the gain pretty minimal, given everything that it does internally?

Reply:
Probably nothing, indeed. The function is probably big enough that it may not qualify for normal inlining, but also does work which certain callers don't care about (and the callers are generally tiny functions) - making it a decent candidate for "LLVM, you should try inlining here, even if you think this function is big enough that you wouldn't otherwise". That said, a bunch of that extra work is probably complicated enough that LLVM wouldn't be able to figure out how to skip it, so...whatever.

Reply:
Got it.  Well, do you think LLVM knows to skip populating a vec if the inlining makes it obvious the vec isn't used?

Reply:
If the contained object didn't have a `Drop` and you gave it enough CPU budget...maybe? I don't think this is worth discussing in this much depth either way, its just one of those things I always figure doing as the "utility function that is big but only called from tiny exposed functions" pattern can be pessimal for the inliner :).